src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -267,34 +305,175 @@ private ValueTask<(HttpConnection, HttpResponseMessage)> GetConnectionAsync(Http
             }
         }
 
+        private async ValueTask<(HttpConnectionBase connection, bool isNewConnection, HttpResponseMessage failureResponse)>
+            GetHttp2ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            Debug.Assert(_kind == HttpConnectionKind.Https || _kind == HttpConnectionKind.SslProxyTunnel);
+
+            // See if we have an HTTP2 connection
+            lock (SyncObj)
+            {
+                if (_http2Connection != null)
+                {
+                    // TODO: Check connection validity and timeouts
+
+                    if (NetEventSource.IsEnabled)
+                    {
+                        Trace("Using existing HTTP2 connection.");
+                    }
+
+                    return (_http2Connection, false, null);
+                }
```

Comment:
Can this be done before the lock as a fast path, e.g. if it's not null, return it, otherwise take the lock and check again?

Reply:
Sure. I reworked this so we only need to take the lock in the case where there's no existing connection, and the semaphore doesn't exist. I.e. it's only protecting creation of the semaphore now. I suppose we could change the semaphore to a Lazy<T>; not sure that matters, thoughts?